### PR TITLE
add option to preserve blanks in text

### PR DESCRIPTION
**Pull request by paed01**
A humble suggestion to solve #1.

### CODE REVIEW

**Comment by paed01** [comment]
The PR doesn't seem to have an obvious impact on performance:

MacBook Pro (15-inch, 2016)
Process: 2,9 GHz Intel Core i7
RAM: 16 GB

PR:
```bash
atlas            : 4.79774 ms/file ± 5.44848
atlas            : 4.65715 ms/file ± 5.41410
atlas            : 4.81922 ms/file ± 5.46177
atlas with option: 4.70645 ms/file ± 5.17285
atlas with option: 4.62679 ms/file ± 5.10887
atlas with option: 4.67465 ms/file ± 5.10796
```

Master:
```bash
atlas: 4.73557 ms/file ± 5.23672
atlas: 4.67620 ms/file ± 5.06736
atlas: 4.45852 ms/file ± 4.99654
```

File: src/HtmlParser.js [hunk 1]
```diff
@@ -3,8 +3,9 @@ const { Transform } = require("stream");
 const { TEXT, NODE, NAME, KEY, VALUE, SCRIPT, STYLE, COMMENT } = require("./states");
 
 module.exports = class HtmlParser extends Transform {
-  constructor(){
+  constructor(options = {preserveNbsp: false}){
```
Comment (atlassubbed) [inline src/HtmlParser.js]: I'm thinking we should treat all whitespace the same -- maybe this should be `preserveWhitespace` instead? Please let me know what you think.

File: src/HtmlParser.js [hunk 2]
```diff
@@ -50,7 +51,9 @@ module.exports = class HtmlParser extends Transform {
       while (i < cacheLen){
         c = cache.charCodeAt(i)
         if (s === TEXT){
-          if (c === 32 || c >= 9 && c <= 13) // ws
+          if ( c === 32 && !preserveNbsp)
```
Comment (atlassubbed) [inline src/HtmlParser.js]: Based on my previous comment, if we treat all whitespace the same, this would change to:

`if (!preserveWhitespace && (c === 32 || c >= 9 && c <= 13)){...}`

and we would get rid of the next `else if` condition

File: test/HtmlParser.test.js [hunk 3]
```diff
@@ -252,4 +252,71 @@ describe("HtmlParser", function(){
...
+        if (data.text) calledData += data.text;
+      });
+      parser.write("Title: ");
+      parser.write("<b> Jan");
+      parser.write(" Bananb");
+      parser.end("erg</b>");
+      expect(calledData).to.equal("Title:  Jan Bananberg");
+    })
+    it("still ignores other whitespace chars when in text", function(){
+      let calledData = "";
```
Comment (atlassubbed) [inline test/HtmlParser.test.js]: If you think this is best, that's fine, too.

**Review by atlassubbed — APPROVED** [review]
I'm going to approve this, but I'll wait to hear your feedback on the `preserveWhitespace` (all whitespace) suggestion before merging. Thanks for your PR 😃

File: src/HtmlParser.js (hunk 2 above)
Comment (paed01) [reply src/HtmlParser.js]: The idea of removing all `\r` and `\n` is really appealing since they *almost* never have an effect on how a browser displays the markup. It compensates for templating frameworks, e.g. handlebars, that keep indentation when rendering html (unless you sprinkle your templates with tildes :().

On the other hand then the lib should treat [pre-tags](https://github.com/paed01/atlas-html-stream/tree/pre-tag) and whatnots that expects unaltered text. It may result in degraded performance.

I guess your suggestion is the way to go!?

Comment (atlassubbed) [reply src/HtmlParser.js]: No, I agree with you, spaces are important so we'll treat them specially. I'll merge this soon, then push `1.1.0` as a minor release. It's 4:00am for me, so I might do this tomorrow.

File: test/HtmlParser.test.js [hunk 4]
```diff
@@ -252,4 +252,71 @@ describe("HtmlParser", function(){
...
+    })
+    it("should still ignore other whitespace chars when in text", function(){
+      let calledData = "";
+      const parser = new HtmlParser({preserveNbsp: true});
+      parser.on("data", data => {
+        if (data.text) calledData += data.text;
+      });
+      parser.end("Title:\r\n<b> Jan \r\n\t Bananberg</b>");
+      expect(calledData).to.equal("Title: Jan  Bananberg");
+    })
```
Comment (atlassubbed) [inline test/HtmlParser.test.js:261]: Looks like the test case fails when non-space whitespace is surrounded by space

Comment (atlassubbed) [reply test/HtmlParser.test.js:261]: Preserving space will be a bit more complicated since the default is to collapse whitespace to a single space, which runs into problems in the above case. 

I think the best option is to perhaps support turning collapse whitespace off (which keeps the parser performant). Remember that this is a stream, so you can always transform your nodes object-by-object outside of this library:

```javascript
// do post-processing on each node in myTransform
const parser = new HtmlParser({preserveWS: true})
myFile.pipe(parser).pipe(myTransform)
```
Where `myTransform` reads the object stream returned by atlas-html-stream. With this option, we won't need to support the `pre` tag because we will return the content as-is anyway. If you only want to collapse outside of the `pre` tag, you could use a `whitespaceCollapser` transform which acts on nodes such that `node.name !== "pre"`, otherwise it returns the content as-is.

Here's the idea behind a `whitespaceCollapser` transform (I just whipped this up, consider it pseudo-code, it is not tested):
```javascript
// myTransform implemented as a whitespace collapse transform
const collapse = text => text.replace(/[\s]/g, " ")
class WhitespaceCollapser extends Transform {
  constructor(){this.inPre = false}
  _transform(node, done){
    if (node.name === "pre") this.inPre = !!node.data;
    else if (node.text != null && !this.inPre)
      node.text = collapse(node.text);
    this.push(node), done(null;
  }
}
```
---

**TLDR** In summary, I think supporting a bunch of special tags and special whitespace collapse may be outside the scope of this library. The great thing about streams is that they can be piped together very easily. I am still interested in providing a `preseveWS` option.

Comment (paed01) [reply test/HtmlParser.test.js:261]: I agree!

File: test/HtmlParser.test.js [hunk 5]
```diff
@@ -252,4 +252,112 @@ describe("HtmlParser", function(){
...
+  describe("preserve whitespace option", () => {
+    theyBothWithPreserveWSOption("should preserve whitespace in raw text", "text", true, function(res, done){
+      expect(res.length).to.equal(1);
+      expect(res[0].text).to.contain("\n");
+      expect(res[0].text).to.contain("  ");
+      done();
+    })
+    theyBothWithPreserveWSOption("should capture all the nodes and whitespaces", "app", true, function(res, done){
+      expect(res.length).to.equal(62)
+      expect(res).to.eql([
```
Comment (atlassubbed) [inline test/HtmlParser.test.js:264]: Cool :) I didn't know this shortcut
